File: CS/SilverlightApplication1/DropController.cs
using System.Collections.Generic;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Shapes;

namespace DevExpress.AgDataGrid.Internal {
	public class DragControllerEx : DragController {
		FrameworkElement _dropContainer;

		public DragControllerEx(FrameworkElement dropContainer, IDragableObject dragableObject, Canvas surfaceCanvas, Point startPt, Point margin)
			: base(dragableObject, surfaceCanvas, startPt, margin) {
			this._dropContainer = dropContainer;
		}

		protected override IDropableObject FindDropableObject(Point point) {
            if(CachedDropableObjects == null)
                BuildDropableObjectsCache();
			FindDropableObjectsToCache(_dropContainer);
            IDropableObject obj = base.FindDropableObject(point); 			
			return obj;
		}
	}

	public class DragObject : IDragableObject {
		AgDataGridRow row;
		AgDataGrid source;
		object dataRow;

		public DragObject(AgDataGridRow row, object dataRow, AgDataGrid source) {
			this.source = source;
			this.row = row;
			this.dataRow = dataRow;
		}

		public AgDataGrid Source { get { return this.source; } }
		public AgDataGridRow Row { get { return this.row; } }
		public object DataRow { get { return this.dataRow; } }

		public FrameworkElement CreateDragShape()
		{
			Rectangle rect = new Rectangle();
			rect.Opacity = 0.8;
			rect.Width = this.row.ActualWidth;
			rect.Height = this.row.ActualHeight;

			LinearGradientBrush gradientBrush = new LinearGradientBrush();

			gradientBrush.StartPoint = new Point(0.5, 1);
			gradientBrush.EndPoint = new Point(0.5, 0);						

			gradientBrush.GradientStops.Add(new GradientStop(){Color=Color.FromArgb(0xFF, 0xBE, 0xDA, 0xF9)});
			gradientBrush.GradientStops.Add(new GradientStop(){Color=Color.FromArgb(0xFF, 0xD3, 0xE6, 0xFB), Offset=1});

			rect.Fill = gradientBrush;
			return rect;
		}
	}

	public delegate bool CanAcceptDelegate(IDragableObject dragObject, Point position);
	public delegate void AcceptDelegate(IDragableObject dragObject, Point position);

	public class DropController : IDropableObject {
		public static bool IsIn(FrameworkElement element, Point localPt) {
			return element != null && localPt.X > 0 && localPt.Y > 0 && localPt.X < element.ActualWidth && localPt.Y < element.ActualHeight;
		}
		
		public static Point TransformSurface(Point pt, object from, object to) {
			UIElement fromElement = from as UIElement;
			UIElement toElement = to as UIElement;
			if(fromElement == null || toElement == null) return pt;
			GeneralTransform gt = toElement.TransformToVisual(fromElement);
			return gt.Transform(pt);
		}

		public DropController(AgDataGrid dataGrid, FrameworkElement dropContainer, Canvas dragSurface) {
			this._dataGrid = dataGrid;
			if(this._dataGrid != null) {
                _dataGrid.DataRowCreated += new DataRowCreatedEventHandler(OnNewDataRow);				
				this._dataGrid.MouseMove += new MouseEventHandler(OnMouseMove);
				this._dataGrid.MouseLeftButtonUp += new MouseButtonEventHandler(OnMouseLeftButtonUp);
			}
			this._dragSurface = dragSurface;
			this._dropContainer = dropContainer;
		}

		AgDataGrid _dataGrid;
		public AgDataGrid DataGrid { get { return this._dataGrid; } }
		Canvas _dragSurface;
		public Canvas DragSurface{
			get { return _dragSurface; }
			set { _dragSurface = value; }
		}
		FrameworkElement _dropContainer;
		public FrameworkElement DropContainer {
			get { return _dropContainer; }
			set { _dropContainer = value; }
		}
		DragController dragController;
		protected DragController DragController { get { return dragController; } }
		public bool IsDragging { get { return DragController != null; } }

		IList<AgDataGridRow> dataRows = new List<AgDataGridRow>();
        void OnNewDataRow(object sender, DataRowCreatedEventArgs e)
        {
			foreach(AgDataGridRow dataRow in dataRows) {
				dataRow.MouseLeftButtonDown -= new MouseButtonEventHandler(OnMouseLeftButtonDown);
			}
			dataRows.Clear();
			AgDataGridRow row = e.Row as AgDataGridRow;
			if(row != null) {
				row.MouseLeftButtonDown += new MouseButtonEventHandler(OnMouseLeftButtonDown);
			}
		}

		void OnMouseLeftButtonUp(object sender, MouseButtonEventArgs e) {
			if(e.Handled) return;
			StopDrag();
		}

		void OnMouseMove(object sender, MouseEventArgs e) {
			if(this.DragSurface == null) return;
			if(DragController != null) { DragController.Move(e.GetPosition(this.DragSurface)); }
		}

		void OnMouseLeftButtonDown(object sender, MouseButtonEventArgs e) {
			if(e.Handled) return;
			StartDragging(sender as AgDataGridRow, e.GetPosition(e.OriginalSource as UIElement));
		}

		protected virtual void StopDrag() {
			if(DragController == null) return;
			DragController.StopDrag();
			this.dragController = null;
			this.DataGrid.ReleaseMouseCapture();
		}


		protected virtual void StartDragging(AgDataGridRow row, Point point) {
			if(this.DataGrid == null) return;
			if(this.DragSurface == null) return;			
			
			this.dragController = new DragControllerEx(
				DropContainer,
				new DragObject(row, DataGrid.GetDataRow(row.Handle), this.DataGrid),
				this.DragSurface, 
				point,
				new Point(this.DataGrid.Margin.Left, this.DataGrid.Margin.Top));
			
			DragController.StartDragShift = AgDataGrid.DefaultStartDragShift;
			DragController.ScrollOnDragging = null;
			this.DataGrid.CaptureMouse(); 
		}

		public event AcceptDelegate DragAccept;
		public event CanAcceptDelegate CanDragAccept;

		#region IDropableObject Members
		void IDropableObject.AcceptDrag(IDragableObject dragObject, Point position) {
			if(DragAccept != null) {
				DragAccept(dragObject, position);
			}
		}

		bool IDropableObject.CanAccept(IDragableObject dragObject, Point position) {
			if(CanDragAccept != null) {
				return CanDragAccept(dragObject, position);
			}
			return false;
		}

		bool IDropableObject.CanAccept() {
			return DragAccept != null && CanDragAccept != null;
		}

		FrameworkElement IDropableObject.GetThumbObject() {
			return null;
		}

		Rect IDropableObject.GetThumbRect(Point position) {
			return new Rect(0, 0, 0, 0);
		}
		#endregion
	}
}

File: VB/SilverlightApplication1/DropController.vb
Imports Microsoft.VisualBasic
Imports System.Collections.Generic
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Input
Imports System.Windows.Media
Imports System.Windows.Shapes

Namespace DevExpress.AgDataGrid.Internal
	Public Class DragControllerEx
		Inherits DragController
		Private _dropContainer As FrameworkElement

		Public Sub New(ByVal dropContainer As FrameworkElement, ByVal dragableObject As IDragableObject, ByVal surfaceCanvas As Canvas, ByVal startPt As Point, ByVal margin As Point)
			MyBase.New(dragableObject, surfaceCanvas, startPt, margin)
			Me._dropContainer = dropContainer
		End Sub

		Protected Overrides Function FindDropableObject(ByVal point As Point) As IDropableObject
			If CachedDropableObjects Is Nothing Then
				BuildDropableObjectsCache()
			End If
			FindDropableObjectsToCache(_dropContainer)
			Dim obj As IDropableObject = MyBase.FindDropableObject(point)
			Return obj
		End Function
	End Class

	Public Class DragObject
		Implements IDragableObject
		Private row_Renamed As AgDataGridRow
		Private source_Renamed As AgDataGrid
		Private dataRow_Renamed As Object

		Public Sub New(ByVal row As AgDataGridRow, ByVal dataRow As Object, ByVal source As AgDataGrid)
			Me.source_Renamed = source
			Me.row_Renamed = row
			Me.dataRow_Renamed = dataRow
		End Sub

		Public ReadOnly Property Source() As AgDataGrid
			Get
				Return Me.source_Renamed
			End Get
		End Property
		Public ReadOnly Property Row() As AgDataGridRow
			Get
				Return Me.row_Renamed
			End Get
		End Property
		Public ReadOnly Property DataRow() As Object
			Get
				Return Me.dataRow_Renamed
			End Get
		End Property

		Public Function CreateDragShape() As FrameworkElement Implements IDragableObject.CreateDragShape
			Dim rect As New Rectangle()
			rect.Opacity = 0.8
			rect.Width = Row.ActualWidth
			rect.Height = Row.ActualHeight

			Dim gradientBrush As New LinearGradientBrush()

			gradientBrush.StartPoint = New Point(0.5, 1)
			gradientBrush.EndPoint = New Point(0.5, 0)

			gradientBrush.GradientStops.Add(New GradientStop() With {.Color = Color.FromArgb(&HFF, &HBE, &HDA, &HF9)})
			gradientBrush.GradientStops.Add(New GradientStop() With {.Color = Color.FromArgb(&HFF, &HD3, &HE6, &HFB), .Offset = 1})

			rect.Fill = gradientBrush
			Return rect
		End Function
	End Class

	Public Delegate Sub CanAcceptDelegate(ByVal dragObject As IDragableObject, ByVal position As Point, ByRef cancel As Boolean)
	Public Delegate Sub AcceptDelegate(ByVal dragObject As IDragableObject, ByVal position As Point)

	Public Class DropController
		Implements IDropableObject
		Public Shared Function IsIn(ByVal element As FrameworkElement, ByVal localPt As Point) As Boolean
			Return element IsNot Nothing AndAlso localPt.X > 0 AndAlso localPt.Y > 0 AndAlso localPt.X < element.ActualWidth AndAlso localPt.Y < element.ActualHeight
		End Function

		Public Shared Function TransformSurface(ByVal pt As Point, ByVal [from] As Object, ByVal [to] As Object) As Point
			Dim fromElement As UIElement = TryCast(from, UIElement)
			Dim toElement As UIElement = TryCast([to], UIElement)
			If fromElement Is Nothing OrElse toElement Is Nothing Then
				Return pt
			End If
			Dim gt As GeneralTransform = toElement.TransformToVisual(fromElement)
			Return gt.Transform(pt)
		End Function

		Public Sub New(ByVal dataGrid As AgDataGrid, ByVal dropContainer As FrameworkElement, ByVal dragSurface As Canvas)
			Me._dataGrid = dataGrid
			If Me._dataGrid IsNot Nothing Then
				AddHandler _dataGrid.DataRowCreated, AddressOf OnNewDataRow
				AddHandler _dataGrid.MouseMove, AddressOf OnMouseMove
				AddHandler _dataGrid.MouseLeftButtonUp, AddressOf OnMouseLeftButtonUp
			End If
			Me._dragSurface = dragSurface
			Me._dropContainer = dropContainer
		End Sub

		Private _dataGrid As AgDataGrid
		Public ReadOnly Property DataGrid() As AgDataGrid
			Get
				Return Me._dataGrid
			End Get
		End Property
		Private _dragSurface As Canvas
		Public Property DragSurface() As Canvas
			Get
				Return _dragSurface
			End Get
			Set(ByVal value As Canvas)
				_dragSurface = value
			End Set
		End Property
		Private _dropContainer As FrameworkElement
		Public Property DropContainer() As FrameworkElement
			Get
				Return _dropContainer
			End Get
			Set(ByVal value As FrameworkElement)
				_dropContainer = value
			End Set
		End Property
		Private dragController_Renamed As DragController
		Protected ReadOnly Property DragController() As DragController
			Get
				Return dragController_Renamed
			End Get
		End Property
		Public ReadOnly Property IsDragging() As Boolean
			Get
				Return DragController IsNot Nothing
			End Get
		End Property

		Private dataRows As IList(Of AgDataGridRow) = New List(Of AgDataGridRow)()
		Private Sub OnNewDataRow(ByVal sender As Object, ByVal e As DataRowCreatedEventArgs)
			For Each dataRow As AgDataGridRow In dataRows
				RemoveHandler dataRow.MouseLeftButtonDown, AddressOf OnMouseLeftButtonDown
			Next dataRow
			dataRows.Clear()
			Dim row As AgDataGridRow = TryCast(e.Row, AgDataGridRow)
			If row IsNot Nothing Then
				AddHandler row.MouseLeftButtonDown, AddressOf OnMouseLeftButtonDown
			End If
		End Sub

		Private Sub OnMouseLeftButtonUp(ByVal sender As Object, ByVal e As MouseButtonEventArgs)
			If e.Handled Then
				Return
			End If
			StopDrag()
		End Sub

		Private Sub OnMouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
			If Me.DragSurface Is Nothing Then
				Return
			End If
			If DragController IsNot Nothing Then
				DragController.Move(e.GetPosition(Me.DragSurface))
			End If
		End Sub

		Private Sub OnMouseLeftButtonDown(ByVal sender As Object, ByVal e As MouseButtonEventArgs)
			If e.Handled Then
				Return
			End If
			StartDragging(TryCast(sender, AgDataGridRow), e.GetPosition(TryCast(e.OriginalSource, UIElement)))
		End Sub

		Protected Overridable Sub StopDrag()
			If DragController Is Nothing Then
				Return
			End If
			DragController.StopDrag()
			Me.dragController_Renamed = Nothing
			Me.DataGrid.ReleaseMouseCapture()
		End Sub


		Protected Overridable Sub StartDragging(ByVal row As AgDataGridRow, ByVal point As Point)
			If Me.DataGrid Is Nothing Then
				Return
			End If
			If Me.DragSurface Is Nothing Then
				Return
			End If

			Me.dragController_Renamed = New DragControllerEx(DropContainer, New DragObject(row, DataGrid.GetDataRow(row.Handle), Me.DataGrid), Me.DragSurface, point, New Point(Me.DataGrid.Margin.Left, Me.DataGrid.Margin.Top))

			DragController.StartDragShift = AgDataGrid.DefaultStartDragShift
			DragController.ScrollOnDragging = Nothing
			Me.DataGrid.CaptureMouse()
		End Sub

		Public Event DragAccept As AcceptDelegate
		Public Event CanDragAccept As CanAcceptDelegate

#Region "IDropableObject Members"
		Private Sub AcceptDrag(ByVal dragObject As IDragableObject, ByVal position As Point) Implements IDropableObject.AcceptDrag
			RaiseEvent DragAccept(dragObject, position)
		End Sub

		Private Function CanAccept(ByVal dragObject As IDragableObject, ByVal position As Point) As Boolean Implements IDropableObject.CanAccept
			If CanDragAcceptEvent IsNot Nothing Then
				Dim cancel As Boolean = False
				CanDragAcceptEvent(dragObject, position, cancel)
				Return cancel
			End If
			Return False
		End Function

		Private Function CanAccept() As Boolean Implements IDropableObject.CanAccept
			Return DragAcceptEvent IsNot Nothing AndAlso CanDragAcceptEvent IsNot Nothing
		End Function

		Private Function GetThumbObject() As FrameworkElement Implements IDropableObject.GetThumbObject
			Return Nothing
		End Function

		Private Function GetThumbRect(ByVal position As Point) As Rect Implements IDropableObject.GetThumbRect
			Return New Rect(0, 0, 0, 0)
		End Function
#End Region
	End Class
End Namespace